### PR TITLE
FIx: Add unique  Program name validation

### LINKED ISSUE #110: Bug: Creation of duplicate organizations throws error
**Describe the bug**
Creation of a duplicate organization throws error and a 500.

**To Reproduce**

As a user, when I try to create an organization it creates successfully, when I try creating another organization with the same payload I see the following error:

Screen Shot 2020-08-16 at 9 23 23 PM

**Expected behavior**
409 for duplicate resource exists should be returned along with appropriate error message

### LINKED ISSUE #205: Bug: Creation of program with non-unique name throws an unhandled error
### Describe the bug
Currently, if we add an organization with the name that was used before (non-unique name), we got an unhandled server error. 
### To Reproduce
1. Go to Swagger UI or client app like Postman/Insomnia
2. Send the  POST request to `/organizations/{organization_id}/programs/program`
3. An unhandled error is returned

### Expected behavior
We should not 500 as a response code with an unprocessed message. We should get handled exception error with the message that the program name is already used.

### Screenshots
image

### Desktop (please complete the following information):
 - OS: elementary OS 5.1.7 Hera (built on Ubuntu)
 - Browser: Google Chrome 
 - Version: 88.0.4324.182

### Additional context
N/A

### PR DESCRIPTION
**Pull request by NenadPantelic**
### Description

Added program name validation. It throws 409 response code if there is already a program with that name. Fixes #205.

### Type of Change:
- Code
- Quality Assurance
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?
1. Through Swagger UI:
image
2. Through unit test -> `test_api_dao_create_program__but_name_not_unique`

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by NenadPantelic** [comment]
@mtreacy002 Can you please review this? I used black formatter so that's the reason why there are so many structuring changes, sorry :sweat_smile:

**Comment by decon-harsh** [comment]
Instead of writing a validation why can't we use , unique =True in Program Name in  Program Model ?

**Comment by NenadPantelic** [comment]
> Instead of writing a validation why can't we use , unique =True in Program Name in Program Model ?

That already exists, but when you try to save the program with the existing name, it throws an error -> exception actually. Validation is part of business logic and the database definitely has its own constraints. There are two ways of handling inappropriate inputs:
1. validate + db constraints
2. just db constraints - some integration exception occurs
Generally, exceptions give too DB specific messages and they're not user-friendly for the clients.  That's one of the reasons why validation is preferred - it will give a message that is specific to some specific case, e.g. name not unique or phone number has not proper format. Of course, you can rely on some DB specific forms of validation that are more flexible, like triggers, but that mixes business logic into DB, then we have an argument about performance trade-off, but this comment is too long even without writing down these details :smile:

**Comment by mtreacy002** [comment]
Thank you, @NenadPantelic for submitting the PR. I'll invite the community members on Zulip to review this PR as I'm a bit low on bandwith atm. Thank you for your understanding.

**Comment by NenadPantelic** [comment]
No prob @mtreacy002 :slightly_smiling_face:

**Review by Rahulm2310 — COMMENTED** [review]
Everything looks fine still there are a lot of unnecessary code formatting changes. It took a lot of time just to review 20 lines of code. I think a PR for #184 would fix this.

**Comment by NenadPantelic** [comment]
> Everything looks fine still there are a lot of unnecessary code formatting changes. It took a lot of time just to review 20 lines of code. I think a PR for #184 would fix this.

I agree, I mentioned that. I saw it late that there was no black formatter in this repo as a standard formatting tool. Thanks for the review :+1:

**Comment by mtreacy002** [comment]
Thanks, @Rahulm2310 for helping with the review. @NenadPantelic now we just need one more reviewer then have it tested.

**Review by devkapilbansal — APPROVED** [review]
LGTM :+1:

File: app/api/validations/organization.py
Comment (ApheleiaS) [inline app/api/validations/organization.py]: No need to add this import, `messages` is already imported.

File: app/api/validations/organization.py
Comment (ApheleiaS) [inline app/api/validations/organization.py]: remove explicit import and add ` return messages.PROGRAM_NAME_IS_MISSING` and `return messages.PROGRAM_NAME_ALREADY_USED`

File: app/api/validations/organization.py
Comment (ApheleiaS) [inline app/api/validations/organization.py]: Same comment as above.

File: app/messages.py
Comment (ApheleiaS) [inline app/messages.py]: Why do these not start with new lines simialr to the change above?

**Review by ApheleiaS — COMMENTED** [review]
Thanks for the PR, left some review comments, please feel free to ask questions.

File: app/messages.py
Comment (devkapilbansal) [reply app/messages.py]: @ApheleiaS some messages are written from new lines because they exceed the length recommended while formatting using a code formatter(black in this case)

File: app/messages.py
Comment (devkapilbansal) [inline app/messages.py]: Only single line is required here

**Review by devkapilbansal — CHANGES_REQUESTED** [review]

File: app/api/validations/organization.py
Comment (devkapilbansal) [reply app/api/validations/organization.py]: Ahh I missed this, thanks @ApheleiaS

Comment (NenadPantelic) [reply app/api/validations/organization.py]: Yeah, I missed that too, I used auto-import from the Pycharm...

**Comment by NenadPantelic** [comment]
Hi @ApheleiaS @devkapilbansal, I made the changes, can you re-review it. Thanks :slightly_smiling_face:

File: app/messages.py
Comment (devkapilbansal) [inline app/messages.py]: Why multiple lines are used here?

File: app/messages.py
Comment (devkapilbansal) [inline app/messages.py]:
```suggestion
    "message": "The website url is not in valid format. It must have http*"
```

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Just some minor changes @NenadPantelic

File: app/messages.py
Comment (NenadPantelic) [reply app/messages.py]: I used black, probably that messed the thing.

**Comment by NenadPantelic** [comment]
@devkapilbansal I reformated `messages.py`. The problem was made by black, I tested it. It creates additional spaces (bounded with double scores) when the file is formatted. Also, if the line is too long (PEP8 suggests 80 chars at most or Pycharm 120 chars -inherited from IntelliJ) it doesn't reformat it, but it does reformat some other lines, maybe it needs some additional flake setup.

File: app/messages.py
Comment (Rahulm2310) [inline app/messages.py]: Correct message description should be `"Phone or mobile fields must be in number format. It must contain numbers and may contain dash "-" or space characters."`

Comment (NenadPantelic) [reply app/messages.py]: @Rahulm2310 They're the same, just reformatted (line 18+19)

File: app/messages.py
Comment (Rahulm2310) [inline app/messages.py]: Add a space between `in` and `the`.

File: app/messages.py
Comment (Rahulm2310) [inline app/messages.py]: I think this message description doesn't sounds good. It should be `"The {field_name} field has to be lesser than {min_limit} characters and not more than {max_limit} characters."`

File: app/messages.py
Comment (Rahulm2310) [inline app/messages.py]: This should be corrected to `only approved fields are submitted`.

File: app/messages.py
Comment (Rahulm2310) [reply app/messages.py]: @NenadPantelic I meant to emphasize the `-` dash.

**Comment by Rahulm2310** [comment]
This PR would have merged long ago, but `black` really made it wait so long.

File: app/messages.py
Comment (NenadPantelic) [reply app/messages.py]: This actually was not part of this issue, but it became silently :joy:

File: app/api/resources/organizations.py
Comment (epicadk) [inline app/api/resources/organizations.py]: This is going to introduce a race condition. Instead you should try to update the program and catch the integrity error that the database throws.

**Review by epicadk — CHANGES_REQUESTED** [review]

File: app/api/resources/organizations.py
Comment (NenadPantelic) [reply app/api/resources/organizations.py]: I agree that generally, this can be prone to a race condition, but it depends on the transactional behavior of the ORM and the isolation level. But yeah I agree that this will be safer if the validation is skipped and the integrity errors are monitored.

Comment (mtreacy002) [reply app/api/resources/organizations.py]: @NenadPantelic , does this mean you will make the changes as suggested by @epicadk?

Comment (NenadPantelic) [reply app/api/resources/organizations.py]: Yes, I will. Exception handling is safer, cause it prevents the possible race condition (between the check and this insert, another process can insert the record).

**Comment by NenadPantelic** [comment]
@mtreacy002 @epicadk @devkapilbansal @Rahulm2310 Changes made. Can you please re-review it? (sorry for so many changes due to formatting with black). Thanks :slightly_smiling_face:

**Review by ApheleiaS — DISMISSED** [review]
LGTM! Thanks for working on this.

**Comment by mtreacy002** [comment]
@NenadPantelic . I tried to run the app but can't get the expected result. Instead, here're the outcome.

Steps:
1. Login as user A who represents an organization, get the detail of the organization user A's representing.
Screen Shot 2021-03-03 at 9 00 37 pm

2. Logout and login as user B who has no organization. 
3. Create an organization for user B with the same name as User A's Organization name.
4. See unexpected error messages:
* on frontend
Screen Shot 2021-03-03 at 9 19 34 pm

* on BIT backend terminal
Screen Shot 2021-03-03 at 9 46 13 pm

see full error in this [gist](https://gist.github.com/mtreacy002/09ce6d8bebd9f2a5ec3e79fe1566411b)

* try to send the load using BIT swagger UI, see unexpected error
Screen Shot 2021-03-03 at 9 28 48 pm
Screen Shot 2021-03-03 at 9 29 12 pm



Can you please investigate why this happens?

**Comment by NenadPantelic** [comment]
@mtreacy002 I will take a look :+1:

**Comment by epicadk** [comment]
It's kind of hard to review the pr because of the large number of changes. Most of them just seem like formatting changes. The is most likely because you haven't called `db.session.rollback()` while handling the integrity error. But I can't say for sure because there are too many changes : /.

**Comment by NenadPantelic** [comment]
I checked it out. Discovered two things:
1) I stashed some changes and forgot to add them to my last commit (rollback more precisely) :joy: 
2) I pulled the changes from my branch and I tested them. It works even without the changes I didn't add. So, it doesn't break, even though the transaction hangs in the air, it is not committed or rollbacked.

So, it works anyway at my local machine, but of course, I committed that last thing - rollback.
P.S. I know the PR is too complex after these formatting changes :cry:

**Review by epicadk — DISMISSED** [review]
Looks good to me. @mtreacy002 can you try now it should work.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@NenadPantelic , I just realised one fundamental thing which is the reason I couldn't get the expected result on the code change you've made. 
Initially, I didn't got a chance to look through the code in detail, but not concerned as it already got view contributors reviewing the PRs. But since I still couldn't get the expected results (even after your latest push), I can't help to wonder and double check. And I found what the culprit is.
What you've provided here on the PR is to add program name validation so it throws 409 response code if there is already a program with that name. 
However, the task you are assigned to on issue #110, is to add **Organization** name validation, not **Program**. 
For now, please open an issue for addressing the bug that you're tackling here (Program name validation check) and link this PR to that issue. I will unassign you from #110 and reassign you to the newly open issue. In hindsight, please take extra care to check what you've been assigned to before start working on the code change 😉
cc @epicadk, @Rahulm2310 , @devkapilbansal and @ApheleiaS

**Comment by NenadPantelic** [comment]
Ah, yikes, sorry for this. I'm currently working on multiple things and I totally got messed up. Sorry :disappointed:. I'll open a new issue and add this PR there, and I'll handle the real problem of this issue. Sorry again :facepalm:

**Comment by mtreacy002** [comment]
@NenadPantelic , please change the issue been fixed here (put on separate line to issue description) and rename the PR title to match the task (Fix: add unique Program name validation)

**Comment by NenadPantelic** [comment]
@mtreacy002  Done :slightly_smiling_face:

**Comment by mtreacy002** [comment]
@rpattath or @faznan3nazer, can one of you please help test this PR? Thanks 😉

File: tests/programs/test_api_create_program.py
Comment (mtreacy002) [inline tests/programs/test_api_create_program.py]: can you please change the test name to "... _when_name_already_exist".

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Well done for the fix, @NenadPantelic . I've tested the fix and it works as expected. Only minor change needed. Please see the comment below.
After you make the change I'll approve and write the test repot. Thanks

File: tests/programs/test_api_create_program.py
Comment (NenadPantelic) [reply tests/programs/test_api_create_program.py]: Changes made

**Comment by NenadPantelic** [comment]
@mtreacy002 I made the requested changes. Please, review it again! Thanks :slightly_smiling_face:

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Sorry, @NenadPantelic , I didn't notice this before, but can you please also exclude the deletion of .github/ISSUE_TEMPLATE/* from your commits? They are not relevant to this PR. I really appreciate your patience on accommodating to the request 🙏.

**Comment by mtreacy002** [comment]
@NenadPantelic , can you please give us update on this?

**Comment by mtreacy002** [comment]
@NenadPantelic , can we please get an update on this?

**Comment by NenadPantelic** [comment]
@mtreacy002 Check it now. Sorry for such a long delay.

**Review by epicadk — DISMISSED** [review]
Race conditions part is solved. Looks good.

File: app/api/resources/organizations.py
Comment (devkapilbansal) [inline app/api/resources/organizations.py]: What's the point of using this?

File: app/api/validations/organization.py
Comment (devkapilbansal) [inline app/api/validations/organization.py:49]: Better to have a blank line to separate both code blocks

File: app/messages.py
Comment (devkapilbansal) [inline app/messages.py:205]: We already have NAME_LENGTH_LESSER_THAN_MAX_LIMIT above. 
```suggestion
NAME_LENGTH_LESSER_THAN_MIN_LIMIT = {
    "message": "The {field_name} has to be at least {min_limit} characters."
```

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Looks good @NenadPantelic
Just some minor changes from my end :point_down:

**Review by epicadk — APPROVED** [review]
👍🏼

**Comment by epicadk** [comment]
I think this is something new in github? @mtreacy002 please approve the pr workflow run.

**Review by mtreacy002 — APPROVED** [review]
LGTM, @NenadPantelic.
Here's the test report:
1. Code review: Done
2. Tested on all possible outcomes:
    Expected behaviour: User should be prevented from creating a program if it's already exist in the data regardless of which organization the program belongs to.
    Actual behaviour: As expected.
    * when the program already exist in their own organization.
    
ezgif com-gif-maker (46)

    * when the program already exist in another organizations.
 
ezgif com-gif-maker (47)


3. Environments:
   * OS: MacOS Catalina 10.15.7
   * Browser: Safari
 
Well done, @NenadPantelic for the fix. Thank you for your contribution 🎉🎉🎉

**Pull request merged**
